Inform blocks of which adjacent blocks triggered an update

ShockedPlot7560 opened the pull request:
## Introduction
Some blocks have a certain cost to check their support during NearbyBlockUpdate. This PR tends to provide a way for blocks to know which sides triggered the update.

### Relevant issues
* Fixes #5982 

## Changes
### API changes
- Deprecated `Block::onNearbyBlockChange()`
- Add `Block::onNearbyBlockChange2(int $flags)` where flags is a set of flags provided by `NearbyBlockChangeFlags`

### Behavioural changes
- Improve performance for some blocks like Cactus

ShockedPlot7560 committed:
base of system

ShockedPlot7560 committed:
start converting blocks

ShockedPlot7560 committed:
bug: fix incorrect checks

ShockedPlot7560 committed:
convert blocks from A to F

JavierLeon9966 commented on src/block/Block.php:
```diff
@@ -454,11 +455,21 @@ public function onBreak(Item $item, ?Player $player = null, array &$returnedItem
+	 * Called when this block or a block immediately adjacent to it changes state.
+	 * @phpstan-param NearbyBlockChangeFlags::FLAG_* $flags
+	 */
+	public function onNearbyBlockChange2(int $flags) : void{
```
Instead of making a new method you should just add the parameter into the already existing one, but add the default flag `0`.

ShockedPlot7560 replied:
Its [break the BC](https://onlinephp.io?s=s7EvyCjg5eLlSs5JLC5WcMrJT85WqObl4iwoTcrJTFZIK81LLsnMz1PIz_NLTSxKqgSrcM5IzEtP1cjMK1FQSctJTC9WsFUw0LRSKMvPTAFp5gTiWl6uWoTBzonJJaXFCqkVJal5KVB7iLUGh8EqEJNtFfJSy6Hma2haAwA%2C&v=8.2.10) and the ambition is to do without. We simply need to find a better name for the function.

JavierLeon9966 commented on src/block/Block.php:
```diff
@@ -454,11 +455,21 @@ public function onBreak(Item $item, ?Player $player = null, array &$returnedItem
 
+	/**
+	 * Called when this block or a block immediately adjacent to it changes state.
+	 * @phpstan-param NearbyBlockChangeFlags::FLAG_* $flags
```
```suggestion
	 * @phpstan-param int-mask-of<NearbyBlockChangeFlags::FLAG_*> $flags
```

ShockedPlot7560 committed:
add utilities function + improve docs

ShockedPlot7560 committed:
finish converting blocks

ShockedPlot7560 committed:
convert all getSides to getSidesAsFacing'

dktapps commented on src/block/Cactus.php:
```diff
@@ -69,9 +69,8 @@ public function onNearbyBlockChange2(int $flags) : void{
-		foreach(NearbyBlockChangeFlags::getSides($flags) as $side){
-			if($side !== NearbyBlockChangeFlags::FLAG_UP){
-				$facing = NearbyBlockChangeFlags::toFacing($side);
+		foreach(NearbyBlockChangeFlags::getSidesAsFacing($flags) as $facing){
```
getFaces() would be fine

dktapps commented on src/block/Bell.php:
```diff
@@ -108,16 +110,19 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
-				break;
+		};
+		foreach(NearbyBlockChangeFlags::getSidesAsFacing($flags) as $supportBlockDirection){
+			if(in_array($supportBlockDirection, $sides, true)){
```
ew no. don't ever use in_array in performance sensitive code

dktapps replied:
This mess can be done better using `containFacing`

dktapps commented on src/block/Block.php:
```diff
@@ -454,11 +455,21 @@ public function onBreak(Item $item, ?Player $player = null, array &$returnedItem
 	/**
 	 * Called when this block or a block immediately adjacent to it changes state.
+	 *
+	 * @deprecated {@see Block::onNearbyBlockChange2()}
```
I'm not sure if this actually works. Better to put the @see on its own line rather than writing it inline.

dktapps reviewed (commented):
I'm not loving this implementation. It's introducing expensive array logic into hot paths.

dktapps commented on src/block/utils/NearbyBlockChangeFlags.php:
```diff
@@ -0,0 +1,137 @@
+	 * @phpstan-param Flag $flag
+	 * @phpstan-param FlagValue ...$others
+	 */
+	public static function contain(int $flag, int ...$others) : bool{
```
This function is unnecessarily costly and also boilerplaty.

e.g.
```php
NearbyBlockChangeFlags::contain($flags, NearbyBlockChangeFlags::FLAG_UP, NearbyBlockChangeFlags::FLAG_DOWN)
```
vs
```php
($flags & (NearbyBlockChangeFlags::FLAG_UP | NearbyBlockChangeFlags::FLAG_DOWN)) !== 0
```

dktapps committed:
Remove mess and clean up code

dktapps committed:
tidy and fix borked glow lichen logic

dktapps committed:
Simplify flag setting mechanism

dktapps commented:
I've made some adjustments and simplified the logic.

StaticSupportTrait could benefit from this, but it would require all of its users to declare which face they are supported by, which is a whole bunch of work. Fortunately, it hasn't made it into a release yet as far as I remember, and SST is already a PITA since it's easy to accidentally look at the wrong block in canBeSupportedAt(), so I was thinking about making a change which would solve this anyway.

dktapps committed:
tidy

ShockedPlot7560 commented on src/block/utils/NearbyBlockChangeFlags.php:
```diff
@@ -0,0 +1,125 @@
+	}
+
+	/**
+	 * @phpstan-param FlagValue $flag
```
Should be `value-of<self::ALL_FACING>`

dktapps replied:
To be honest I'm not sure if this function should reject FLAG_SELF. It's easy to see how this could result in crashes, and before my amendments there were a couple of cases that could have plausibly crashed.

ShockedPlot7560 commented on src/world/World.php:
```diff
@@ -299,8 +300,8 @@ class World implements ChunkManager{
 	/**
-	 * @var true[] blockhash => dummy
-	 * @phpstan-var array<BlockPosHash, true>
+	 * @var int[] blockhash => dummy
```
```suggestion
	 * @var int[] blockhash => update reference block flags
```

dktapps replied:
where does "reference" come into it?

ShockedPlot7560 replied:
Its mean the blocks referent for the update, no?

dktapps replied:
That's a very confusing way of putting it. "cause flags" would be simpler.

dktapps committed:
World: fix documentation

dktapps committed:
NearbyBlockChangeFlags: reduce API surface

dktapps committed:
NearbyBlockChangeFlags: remove unnecessary prefixes

dktapps committed:
NearbyBlockChangeFlags: do not assign values based on other flags
this made it very difficult to see what was going on.
it would be ideal if we could do something like golang's `1 << iota` macro, but sadly that's not an option here.

dktapps committed:
Improve rails

dktapps committed:
tidy

dktapps committed:
Vine: fixed update logic
since vine faces can have side, up, or both supports, all faces must be rechecked if the upper block changed, regardless of whether the side blocks changed or not.
If only the side blocks changed, then the original logic applies.

dktapps committed:
Vine: use existing variable

dktapps committed:
Improve face extraction for cactus and concrete powder

dktapps committed:
DoublePlant: fixed self-destruct logic
this wasn't correctly accounting for halves of the plant being replaced by other blocks, leading to the bottom half persisting when it shouldn't.

dktapps committed:
Fire: check all sides always
since we don't know which neighbour provided the block with support (or if multiple blocks did), we have to recheck all of them, regardless of which one changed, as we need to know whether there are _any_ support blocks remaining.

dktapps committed:
NearbyBlockChangeFlags: optimize for the 99% case
it's rare that we need to check multiple faces at once (this case only comes up once in the current code), so it doesn't make sense for all the other usages to pay a performance penalty for this functionality.
the 1% cases can just call hasFace() multiple times anyway, which is what fence gates are now doing.

dktapps committed:
Fixed doc comment

dktapps committed:
Cactus: added missing break

dktapps committed:
Fixed self updates not working correctly

when a self update occurs, it's usually because the block itself has just been set. This might be because of growth, placement, or a variety of other factors.
in such cases, all surrounding checks must be done to ensure correct behaviour.
in essence, a block that was just created has had all of its neighbours changed at once, so this is kinda elegant.
this also removes the need for a dedicated self flag, which was only checked when needing to workaround this bug anyway.

dktapps committed:
Fixed some doc comments

dktapps committed:
Added some TODO comments

dktapps committed:
Update PHPStan baselines

dktapps committed:
Fix CS

ShockedPlot7560 commented on src/block/BaseRail.php:
```diff
@@ -220,17 +221,24 @@ private function setConnections(array $connections) : void{
+		if(($flags & NearbyBlockChangeFlags::DOWN) !== 0 && !$this->getAdjacentSupportType(Facing::DOWN)->hasEdgeSupport()){
+			$break = true;
+		}
+		foreach($this->getCurrentShapeConnections() as $connection){
```
We will enter the loop even if the break is true

ShockedPlot7560 commented on src/block/Cactus.php:
```diff
@@ -61,6 +62,21 @@ public function onEntityInside(Entity $entity) : bool{
+		if(($flags & NearbyBlockChangeFlags::DOWN) !== 0){
+			$supportBlock = $this->getSide(Facing::DOWN);
+			if(!$supportBlock->hasSameTypeId($this) && !$supportBlock->hasTypeTag(BlockTypeTags::SAND)){
+				$this->position->getWorld()->useBreakOn($this->position);
```
Could be set a new `break` var to avoid entering in the loop later?

dktapps replied:
Maybe. This whole self-destruct thing needs some cleanup work tbh, so I wouldn't think too much about it.

dktapps committed:
Integrate flags with StaticSupportTrait

dktapps commented:
Performance testing suggests that the benefit of this is smaller than I hoped, because we're optimizing the wrong thing.

Local testing with large wheat farms showed that each block update takes about 4.2 microseconds total without the PR. However, the time spent in onNearbyBlockChange() is only about 1.4 microseconds (at least for wheat crops and farmland).

This PR reduces the total to 3.5 us with an average of 0.7 us spent in onNearbyBlockChange2() after integrating with StaticSupportTrait. It's not bad (15-20% performance improvement), but it's not the big win I was hoping for.

It seems like there needs to be more work done to:
- reduce and optimize work done in World->actuallyDoTick() for block updates
- reduce the actual number of block updates (difficult to do without breaking observers)

There are major performance gains to be had in crop farms, where the vast majority of the CPU time spent on crops is actually coming from the useless updates rather than the random growth ticks.

dktapps commented:
A new comparison profile needs to be done with https://github.com/pmmp/PocketMine-MP/commit/6a3ec70c72e79603b2cc618269461204fd578f2b. I suspect that this commit closes the performance gap by quite a bit.

dktapps committed:
Merge branch 'minor-next' into feat/neighboor-block-update

dktapps committed:
Merge branch 'minor-next' into feat/neighboor-block-update

dktapps commented:
> A new comparison profile needs to be done with [6a3ec70](https://github.com/pmmp/PocketMine-MP/commit/6a3ec70c72e79603b2cc618269461204fd578f2b). I suspect that this commit closes the performance gap by quite a bit.

Now shows ~7% performance improvement (after optimizations to `Block->getSide()`)
Before: https://timings.pmmp.io/?id=312959
After: https://timings.pmmp.io/?id=312960

Everything is within margin of error, except the Scheduled Block Updates timing, which shows an overall reduction from 7.44 m to 6.94 m (I ignore the average time per tick since it includes ticks where nothing happened).

ShockedPlot7560 committed:
Merge remote-tracking branch 'upstream/minor-next' into feat/neighboor-block-update

ShockedPlot7560 committed:
fix loop breaking path

dktapps commented:
To paraphrase #internals-dev discussions, I'm unconvinced that this much extra code is justified by a 7% performance increase, particularly since this is a minor part of the problem anyway. I'm going to leave this open for now, but it seems like other ways to improve performance need to be explored.

dktapps commented:
Closing this as the aforementioned minimal performance gains don't seem to justify the large amount of extra code to maintain.
